src/fde.h

```diff
@@ -97,6 +112,12 @@ class fde
     /// record a transaction on this FD
     void noteUse() { ++pconn.uses; }
 
+    /// record bytes being read and return the accumulated total for this FD.
+    uint64_t bytesRead(size_t n=0) { if (n > 0) bytes_read += n; return bytes_read; }
```

Comment:
Please split this method into two, one to increment the counter and one to get the total. The split will simplify (and might even speed up) both methods!

* I suggest naming the first method noteBytesRead() or recordBytesRead(), but there may be better names.
* I would keep the name of the constant getter method the same -- bytesRead(), but totalBytesRead() also works OK IMO.
* To simplify searches and references, I would _not_ use the same name for both methods.

In the first method -- noteBytesRead() -- please do not waste cycles/ink on the positivity check: Zeros are rare and harmless in this context, while the `> 0` check may create a false impression that negative arguments are OK (when, in fact, they are caller bugs).

Same for bytesWritten(), of course.

Reply:
Zeros may be harmless but negatives are not, and they are relatively common. See the original fd_bytes() implementation for comparision. Placing the if-condition inside this function de-duplicates all callers having to perform the identical check themselves prior to calling it.

Reply:
> Zeros may be harmless but negatives are not, and they are relatively common. See the original fd_bytes() implementation for comparision. Placing the if-condition inside this function de-duplicates all callers having to perform the identical check themselves prior to calling it.

Isn't size_t guaranteed to be unsigned?

Reply:
> Isn't size_t guaranteed to be unsigned?

Yes, it [is](https://en.cppreference.com/w/c/types/size_t), and the proposed methods should continue to use `size_t` (not `int` that was used by fd_bytes()).

And, yes, the callers should check the I/O result before interpreting it as the number of bytes read or written. Those checks are a part of the I/O APIs they are using.

Reply:
Ah, that is what this is all about. Thank you @kinkie for pointing out the real problem.

Reply:
fixed in [bb0ee01](https://github.com/squid-cache/squid/pull/1352/commits/bb0ee01abadf8c307baaff54043962e528ef2c2e)